openedx_events/learning/signals.py

```diff
@@ -196,3 +197,58 @@
         "notification_data": UserNotificationData,
     }
 )
+
+# .. event_type: org.openedx.learning.exam.attempt.submitted.v1
+# .. event_name: EXAM_ATTEMPT_SUBMITTED
+# .. event_description: Emitted when an exam attempt is submitted by a learner.
```

Comment:
Some high-level questions:
1. Does the platform have other types of "exams" that would not trigger these events? If so, how do we differentiate? Note that an event is defined by its event trigger + event data.
2. For the exams covered by these events, are there multiple exam types? Would it be useful to know that as part of the event data?
3. If there are different exam types, do some of these events only apply to certain types of exam types? If so, would that be useful to note?

Reply:
1. kind of. The existing exam system (edx-proctoring) does not trigger these events, the service that emits these is intended to replacement. There are no other types of exams outside that.
2. Good point. There are multiple exam types and edx-platform is _mostly_ unaware of this since all the state/ui flows are manage by the exams/proctoring service. That said, I think there is value including this. Consumers may likely care if an exam is proctored or not. We might have actually missed a need for this. I think Credits depend on if an exam is proctored.
3. Yup some of these events are impossible depending on the exam type. A timed exam cannot be rejected since there is no video review to reject for example. We should add those details to the event descriptions

Reply:
Are there non-timed exams, and are those included in this as well? Or are these just timed-exams, proctored and not-proctored? Just curious.

Reply:
🤔  I guess this doesn't include 'all' exams just 'special exams' which have a different UX and workflow. Right now, all special exam types are a subset of timed. This wouldn't include normal sections instructors are using as or are graded as exams however. We could possibly reintroduce the term 'special exam' for these events which could clear this up.

Reply:
If you think of this more as an API discussion around exams and special-exams, who do you think should be a part of this discussion? Maybe https://openedx.atlassian.net/wiki/spaces/COMM/pages/3554214289/Arch+Coordination+-+How+to+announce+and+tune-in will help you get the right discussion going?

Reply:
Related: There are suggestions https://docs.openedx.org/projects/openedx-events/en/latest/how-tos/creating-new-events.html#propose-the-new-event-to-the-community for how to propose these new events.

Reply:
Hi Robert!

To answer your question, you could technically title a course subsection as an "Exam" and grade it as an exam, however our exams backend (i.e. the exams IDA we're developing, edx-exams) would never know about it, as the exam backend is only used/called when a subsection is marked as a "Special Exam", i.e. a timed or proctored exam.

As such, the events that we are defining would only ever be produced in the context of a timed or proctored exam. The logic that we're implementing to produce these events mirrors the logic that edx-proctoring uses to call functions in various edx-platform services (e.g. calling the credit service to set a leaner's credit requirement status).

The bulk of the logic that we're porting from edx-proctoring can be seen here: https://github.com/openedx/edx-proctoring/blob/8091977772ec8967c79e854de7e0c813ee7c3d7e/edx_proctoring/api.py#L1479 (I recommend viewing this in an IDE since it can go a little deep).

Since certain exam attempt statuses (i.e. verified, rejected, and errored) are inherently linked to proctored exams, this means that edx-proctoring can distinguish if an exam is proctored or not based on the incoming status to update the attempt to (Side note that we never have a case where we need to know if an exam is specifically timed and not proctored).

Since we are simply mirroring this behavior in the exams IDA, this means that events produced in the exams IDA would be consumed by the intended services naturally, without us needing to gate anything behind a if exam_type=='proctored' kind of statement.

The pseudocode of how we'll mimic edx-proctoring's logic in edx-exams can be seen in an internal document that I can DM to you.

With all this being said, I agree that there could be value in attaching the exam type to the event down the line.

Reply:
> We could possibly reintroduce the term 'special exam' for these events which could clear this up.

Something like this sounds reasonable, but the point of my last two messages is that it makes sense to bring in some other domain experts to help decide this type of question.

Reply:
@ilee2u: I posted my last comment at the same time you were posting yours, but it applies equally to you. This all sounds reasonable, but it would be great to get others involved who work in this area to help with names and details to make the events more clear and forward thinking.

Reply:
@ilee2u I think a discuss post referencing this PR in combination with the [original ADR](https://github.com/openedx/edx-proctoring/pull/1049) for the project would be a good start to get input

Reply:
A Discuss post for this topic has been created here: https://discuss.openedx.org/t/how-to-implement-events-for-the-exams-ida/11276

Reply:
Can you confirm that you agreed/decided that no name or doc changes were needed to clarify what is meant by "exam"? It feels like at least a doc update might be warranted, unless you think I'm the only person that would ever ask about this.

Reply:
Hello folks. I do believe we should use a different name for the event. `exam` might mean something in `edx-exams`, but it could mean other (similar) things in `edx-platform` that need another type of data for the event. What do you think?

Reply:
^ I think this is a good idea to have before we merge. I could see people getting confused on two points

1. these events only pertain to 'special' exams
2. these events are not published by the existing exams system

Reply:
I've made the following changes to address this:

- Changed the data name to "SpecialExamAttemptData"
- Added a comment about how events using this data are only emitted for special exams from the newer exams backend, edx-exams

Reply:
That's better! Thank you folks.

Reply:
I don't want to push hard on this, because it's not my domain of expertise, but the naming difference between EXAM_ATTEMPT_SUBMITTED and SpecialExamAttemptData just catches my attention. The obvious other option is to rename to SPECIAL_EXAM_ATTEMPT_SUBMITTED , and I guess the question I have is around the pros/cons of going with the more specific name.

Reply:
🤔 🤔 It's a good question, I'd almost lean towards not using 'special' in the name at all and just commenting to the meaning? Otherwise, we need to consistently use the term across Open edX. I don't think 'special exam' has any additional meaning to anyone working outside this domain unless it's described. There was a actually discussion a while back that we stop using this term because it's just confusing, trying to dig that up.

I'm not sure how to best describe the fact that this intends to include all content that is treated with an exam user experience, but not content that might be titled (on the section or grading policy) by course teams as 'exam'.  The latter is ultimately no different from other content, the platform doesn't do anything functionally different. (hence not special)